tokio-postgres/src/copy_both.rs

```diff
@@ -0,0 +1,345 @@
+        stream_sender: mpsc::Sender<Result<Bytes, Error>>,
+    ) -> CopyBothReceiver {
+        CopyBothReceiver {
+            message_receiver,
+            sink_receiver,
+            stream_sender,
+            state: CopyBothState::CopyBoth,
+            buffered_message: None,
+        }
+    }
+
+    /// Convenience method to set the subprotocol into an unexpected message state
+    fn unexpected_message(&mut self) {
+        self.sink_receiver.close();
+        self.buffered_message = Some(Err(Error::unexpected_message()));
+        self.state = CopyBothState::CommandComplete;
+    }
+
+    /// Processes messages from the backend, it will resolve once all backend messages have been
+    /// processed
+    fn poll_backend(&mut self, cx: &mut Context<'_>) -> Poll<()> {
+        use CopyBothState::*;
+
+        loop {
+            // Deliver the buffered message (if any) to the user to ensure we can potentially
+            // buffer a new one in response to a server message
+            if self.buffered_message.is_some() {
+                // If the receiver is gone we'll just drop the message
+                let _ = ready!(self.stream_sender.poll_ready(cx));
+                let message = self.buffered_message.take().unwrap();
+                let _ = self.stream_sender.start_send(message);
+            }
+
+            match ready!(self.message_receiver.poll_next_unpin(cx)) {
+                Some(Ok(Message::CopyData(body))) => match self.state {
+                    CopyBoth | CopyOut => {
+                        self.buffered_message = Some(Ok(body.into_bytes()));
+                    }
+                    _ => self.unexpected_message(),
+                },
+                // The server->client stream is done
+                Some(Ok(Message::CopyDone)) => {
+                    match self.state {
+                        CopyBoth => self.state = CopyIn,
+                        CopyOut => self.state = CopyNone,
+                        _ => self.unexpected_message(),
+                    };
+                }
+                Some(Ok(Message::CommandComplete(_))) => {
+                    match self.state {
+                        CopyNone => self.state = CopyComplete,
+                        CopyComplete => {
+                            self.stream_sender.close_channel();
+                            self.sink_receiver.close();
+                            self.state = CommandComplete;
+                        }
+                        _ => self.unexpected_message(),
+                    };
+                }
+                // The server indicated an error, terminate our side if we haven't already
+                Some(Err(err)) => {
+                    match self.state {
+                        CopyBoth | CopyOut | CopyIn => {
+                            self.sink_receiver.close();
+                            self.buffered_message = Some(Err(err));
+                            self.state = CommandComplete;
```

Comment:
No because we're still waiting for the `ReadyForQuery` message and we also need to loop around and start the sending of the buffered error to the user.